File: backend/app/domain/driver_location.py
"""Latest authorized trip position; no historical GPS trail is retained."""

from dataclasses import dataclass
from datetime import datetime
from uuid import UUID

from app.domain.entities import RideStatus

TRACKED_RIDE_STATUSES = {RideStatus.ACCEPTED, RideStatus.ARRIVING, RideStatus.IN_PROGRESS}
LOCATION_RETENTION_SECONDS = 120


@dataclass(frozen=True)
class DriverLocation:
    ride_id: UUID
    driver_id: UUID
    latitude: float
    longitude: float
    accuracy_meters: float
    heading: float | None
    captured_at: datetime
    received_at: datetime

File: mobile/src/features/tracking/domain/driverLocation.ts
import type { Ride } from '@/features/rides/domain/types';

export type DriverLocation = {
  rideId: string;
  driverId: string;
  latitude: number;
  longitude: number;
  accuracyMeters: number;
  heading: number | null;
  capturedAt: string;
  receivedAt: string;
};

export function canTrackRide(ride: Pick<Ride, 'status'> | null | undefined): boolean {
  return !!ride && ['accepted', 'arriving', 'in_progress'].includes(ride.status);
}

export function newerLocation(current: DriverLocation | null | undefined, incoming: DriverLocation | null) {
  if (!incoming) return null;
  if (!current || current.rideId !== incoming.rideId || current.driverId !== incoming.driverId) return incoming;
  return Date.parse(incoming.capturedAt) > Date.parse(current.capturedAt) ? incoming : current;
}

export function locationFreshness(location: DriverLocation | null, now = Date.now()) {
  if (!location) return 'waiting' as const;
  const age = now - Date.parse(location.capturedAt);
  if (age > 120_000 || age < -10_000) return 'unavailable' as const;
  return age > 20_000 ? 'stale' as const : 'live' as const;
}
